physicalplan/aggregate/hashaggregate.go

```diff
@@ -152,7 +152,11 @@ func (a *aggregate) initializeTables(ctx context.Context) error {
 }
 
 func (a *aggregate) workerTask(workerID int, vector model.StepVector) model.StepVector {
-	table := a.tables[workerID]
-	table.aggregate(vector)
-	return table.toVector(a.vectorPool)
+	if len(a.tables) > workerID {
```

Comment:
Hm, what happens if this condition is false? Does it mean there's nothing to evaluate? I wonder if the edge case should be prevented somewhere before calling this method 🤔

Reply:
So taking a deeper look here. For the HashAggregate operator, this panic only occurs when `Next()` is called before `Series()` with empty data. This was detected from instant queries as the `Exec` method calls them in this order but range query `Exec` does not.

So when we call `Next()` before `Series()`, we have a data race, as some of the workers have started executing their tasks by reading from the aggregate tables slice, but `initializeTables` has not been called yet. The `if` here, in this case, is actually wrong.

Should `Series()` always be called before `Next()`?

Reply:
Hm, but don't we initialize tables [before](https://github.com/thanos-community/promql-engine/blob/main/physicalplan/aggregate/hashaggregate.go#L86) we start [sending](https://github.com/thanos-community/promql-engine/blob/main/physicalplan/aggregate/hashaggregate.go#L94) tasks to workers?

Reply:
Yeah I thought so too, but the panic points to the table array being read in `workerTask` as out of range. There might be some race within `worker.Send` and `worker.Start` (this seems to execute the `workerTask` via `doWork`). I'll try to look into this a bit more today

Reply:
Ok so, it was actually the `Shutdown` method. It releases the `next` lock which allows `Start` to acquire it and execute `doWork` i.e `aggregate.workerTask`. In hashaggregate `Next()` if we don't get data from `a.next.Next()`  we call `Shutdown` on workers [here](https://github.com/thanos-community/promql-engine/blob/main/physicalplan/aggregate/hashaggregate.go#L81), which happened in case of empty series, and this allowed `Start` to execute `doWork` and cause panics. I fixed this, and also added comments about how the locks are used for syncing across worker methods! 🙂